client-js/main/client/composite-client.js

```diff
@@ -0,0 +1,232 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import ObjectToProto from "./object-to-proto";
+import {CommandHandlingError, CommandValidationError, SpineError} from "./errors";
+import {Status} from '../proto/spine/core/response_pb';
+import {Client} from "./client";
+
+/**
+ * A {@link Client} delegates requests to case-specific client implementations.
+ *
+ * @see QueryingClient
+ * @see SubscribingClient
+ * @see CommandingClient
+ */
+export class CompositeClient extends Client {
+
+    constructor(querying, subscribing, commanding) {
+        super();
+
+        /**
+         * @type QueryingClient
+         *
+         * @private
+         */
+        this._querying = querying;
+
+        /**
+         * @type SubscribingClient
+         *
+         * @private
+         */
+        this._subscribing = subscribing;
+
+        /**
+         * @type CommandingClient
+         *
+         * @private
+         */
+        this._commanding = commanding;
+    }
+
+    newQuery() {
+        return this._querying.newQuery();
+    }
+
+    execute(query) {
+        return this._querying.execute(query);
+    }
+
+    fetch({entity: cls, byIds: ids}) {
+        return this._querying.fetch(cls, ids);
+    }
+
+    newTopic() {
+        return this._subscribing.newTopic();
+    }
+
+    subscribeTo(topic) {
+        return this._subscribing.subscribeTo(topic);
+    }
+
+    subscribe({entity: cls, byIds: ids}) {
+        return this._subscribing.subscribe(cls, ids);
+    }
+
+    sendCommand(commandMessage, acknowledgedCallback, errorCallback, rejectionCallback) {
+        return this._commanding.sendCommand(commandMessage,
+                                            acknowledgedCallback,
+                                            errorCallback,
+                                            rejectionCallback);
+    }
+}
+
+/**
+ * A client which performs entity state queries.
+ */
+export class QueryingClient {
+
+    /**
+     * @param {!ActorRequestFactory} actorRequestFactory
+     *        a request factory to build requests to Spine server
+     */
+    constructor(actorRequestFactory) {
+        this._requestFactory = actorRequestFactory;
+    }
+
+    /**
+     * Executes the given `Query` instance specifying the data to be retrieved from
+     * Spine server fulfilling a returned promise with an array of received objects.
+     *
+     * @param {!Query} query a query instance to be executed
+     * @return {Promise<<T extends Message>[]>} a promise to be fulfilled with a list of Protobuf
+     *        messages of a given type or with an empty list if no entities matching given query
+     *        were found; rejected with a `SpineError` if error occurs;
+     *
+     * @template <T> a Protobuf type of entities being the target of a query
+     */
+    execute(query) {
+        throw new Error('Not implemented in abstract base.');
```

Comment:
In our JS code in some places we mark abstract classes and methods with `@abstract`, maybe we could benefit from it in this changeset.

Especially for classes, it would be convenient to see the class is abstract right away.